docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,573 @@
+extending the infra container to reap orphaned zombies, and [rolling
+out](https://pr.k8s.io/37404).
+
+#### Pod Level cgroup
+
+Tracked in [#26751](https://issues.k8s.io/26751), Kubernetes currently enforces
+resource constraints at a container level. A container added by *Running Debug*
+cannot allocate its own resources and must instead fit within previously
+allocated pod resources.
+
+A pod could potentially be evicted during debugging, particularly if there are
+insufficient resources to allow for the troubleshooting process, but there's not
+much we can do about this until Kubernetes adds support for vertical pod
+resource scaling ([#10782](https://issues.k8s.io/10782)).
+
+### Implementation Plan
+
+#### 1.7 Alpha Release
+
+We're targeting an alpha release in Kubernetes 1.7 that includes the following
+basic functionality:
+
+*   Support in the kubelet for attaching debug containers to a running pod
+*   A `kubectl debug` command to initiate a debug container
+*   `kubectl describe pod` will list status of debug containers running in a pod
+
+Functionality in the kubelet will be hidden behind an alpha feature flag and
+disabled by default. The following are explicitly out of scope for the 1.7 alpha
+release:
+
+*   *Copy Debug* will not be implemented
+*   `kubectl describe pod` will not report the running command, only the
+    container status
+*   Garbage collection of exitted debug containers cannot be tuned
+*   Specific integration with cluster admission controller and pod security
+    policy, though these may work by default
+
+*Running Debug* is implemented in the kubelet's generic runtime manager.
+Performing this operation with a legacy (non-CRI) runtime or a cluster without
+the alpha feature enabled will result in a not implemented error. Implementation
+will be broken into the following steps:
+
+##### Step 1: Container Type
+
+Debug containers exist within the kubelet as a `ContainerStatus` without a
+container spec. As with other `kubecontainers`, their state is persisted in the
+runtime via labels.
+
+To distinguish debug containers, the runtime uses a new
+`io.kubernetes.container.type` label. Existing containers will be started with a
+type of `REGULAR`, `INIT` or `SANDBOX`. When added in a subsequent step, debug
+containers will start with with the type `DEBUG`. The `type` label will populate
+a new field `Type` in `container.ContainerStatus`.
+
+##### Step 2: Creation and Handling of Debug Containers
+
+This step adds methods for creating debug containers, but doesn't yet modify the
+kubelet API. The kubelet will gain a `RunDebugContainer()` method which accepts
+a `v1.Container` and creates a debug container. Fields that don't make sense in
+the context of a debug container (e.g. probes, lifecycle, resources) will be
+excluded by a white list and ignored.
+
+The kubelet will treat `DEBUG` containers differently in the following ways:
+
+1.  `SyncPod()` ignores containers of type `DEBUG`, since there is no
```

Comment:
But the proposal said `SyncPod` would just ignore the debug containers.... It can't really ignore AND compute changes for the debug containers. I'm a little bit concerned about the increased complexity of the code which may *at times* ignore the debug container. I know this is the implementation detail, but please make sure the logic clear and readable in the code.